In [141]:
#This line configures the plotting resolution for retina screens

%config InlineBackend.figure_format = 'retina'

In [142]:
cd /Users/zacklarsen/Desktop/CSC 575/HW/HW2

/Users/zacklarsen/Desktop/CSC 575/HW/HW2


In [143]:
from __future__ import division   #Importing this because otherwise there is no precision after decimal place

In [144]:
#This is really nice. It lets you print multiple variables without print statements

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [145]:
import pylab as pl
import numpy as np
np.set_printoptions(precision=8)
from numpy import linalg as la
import pandas as pd
from sklearn.metrics import confusion_matrix, precision_score, recall_score

# 1 Retrieval from an inverted index

### Please see the word document for the answers to question 1. What follows below is just for certain calculations.

In [146]:
# This index is provided in the homework handout, and I am simply creating a dataframe out of it below:

index = {'Document' : pd.Series(('Doc1','Doc2','Doc3')),
    'allow' : pd.Series((1,0,0)),
    'base' : pd.Series((1,0,0)),
    'cluster' : pd.Series((0,0,2)),
    'collect' : pd.Series((1,0,0)),
    'creat' : pd.Series((0,0,2)),
    'critic' : pd.Series((0,1,0)),
    'default' : pd.Series((1,0,0)),
    'document' : pd.Series((1,1,2)),
    'effici' : pd.Series((0,1,0)), 
    'engine' : pd.Series((2,0,1)),
    'evalu' : pd.Series((0,1,0)),
    'feedback' : pd.Series((0,1,0)),
    'file' : pd.Series((1,0,0)),
    'full' : pd.Series((0,0,1)),
    'glimps' : pd.Series((2,0,0)),
    'good' : pd.Series((0,0,1)),
    'index' : pd.Series((1,2,0)),
    'inform' : pd.Series((1,0,0)),
    'larg' : pd.Series((0,1,0)),
    'larger' : pd.Series((1,0,0)),
    'main' : pd.Series((0,1,0)),
    'part' : pd.Series((1,0,0)),
    'process' : pd.Series((0,2,0)),
    'queri' : pd.Series((1,2,0)),
    'quickli' : pd.Series((1,0,0)),
    'relev' : pd.Series((0,1,0)),
    'retriev' : pd.Series((1,1,1)),
    'scale' : pd.Series((0,1,0)),
    'search' : pd.Series((3,0,1)),
    'short' : pd.Series((0,0,1)),
    'snippet' : pd.Series((0,0,1)),
    'system' : pd.Series((3,2,0)),
    'text' : pd.Series((0,0,1)),
    'updat' : pd.Series((0,1,0)),
    'us' : pd.Series((1,0,0)),
    'web' : pd.Series((1,0,2)),
    }

index = pd.DataFrame(data=index)
index = index.set_index('Document')
index
index.T.head(6)

allow  base  cluster  collect  creat  critic  default  document  \
Document                                                                    
Doc1          1     1        0        1      0       0        1         1   
Doc2          0     0        0        0      0       1        0         1   
Doc3          0     0        2        0      2       0        0         2   

          effici  engine ...   retriev  scale  search  short  snippet  system  \
Document                 ...                                                    
Doc1           0       2 ...         1      0       3      0        0       3   
Doc2           1       0 ...         1      1       0      0        0       2   
Doc3           0       1 ...         1      0       1      1        1       0   

          text  updat  us  web  
Document                        
Doc1         0      0   1    1  
Doc2         0      1   0    0  
Doc3         1      0   0    2  

[3 rows x 36 columns]

Document  Doc1  Doc2  Doc3
allow        1     0     0
base         1     0     0
cluster      0     0     2
collect      1     0     0
creat        0     0     2
critic       0     1     0

In [147]:
DF = np.array([(index!=0).sum(0)]) 
DF

array([[1, 1, 1, 1, 1, 1, 1, 3, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 3, 1, 2, 1, 1, 2, 1, 1, 1, 2]])

In [148]:
N = len(index.T)
N

36

In [149]:
DF

array([[1, 1, 1, 1, 1, 1, 1, 3, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 3, 1, 2, 1, 1, 2, 1, 1, 1, 2]])

In [188]:
# Calculating IDF values for each term


IDF = np.log2((np.divide(N, DF)))

IDF

array([[ 5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  3.5849625,  5.169925 ,
         4.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,
         5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  3.5849625,
         5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ]])

In [182]:
colnames = list(index)
colnames[0:5]
#colnames = pd.DataFrame(colnames)

['allow', 'base', 'cluster', 'collect', 'creat']

In [186]:
index['allow'][0]

1

In [185]:
index[colnames[0]][0]

1

In [189]:
IDF

array([[ 5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  3.5849625,  5.169925 ,
         4.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,
         5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  3.5849625,
         5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  4.169925 ,  5.169925 ,  5.169925 ,  5.169925 ,  4.169925 ]])

In [196]:
IDF[0][1]

5.1699250014423122

In [219]:
# Construct a dictionary for the terms and their IDF weights

term_Dict = {}

for i in range(0,36):
    #print i
    term = colnames[i]
    term_Dict[term] = IDF[0][i]
    
term_Dict

{'allow': 5.1699250014423122,
 'base': 5.1699250014423122,
 'cluster': 5.1699250014423122,
 'collect': 5.1699250014423122,
 'creat': 5.1699250014423122,
 'critic': 5.1699250014423122,
 'default': 5.1699250014423122,
 'document': 3.5849625007211561,
 'effici': 5.1699250014423122,
 'engine': 4.1699250014423122,
 'evalu': 5.1699250014423122,
 'feedback': 5.1699250014423122,
 'file': 5.1699250014423122,
 'full': 5.1699250014423122,
 'glimps': 5.1699250014423122,
 'good': 5.1699250014423122,
 'index': 4.1699250014423122,
 'inform': 5.1699250014423122,
 'larg': 5.1699250014423122,
 'larger': 5.1699250014423122,
 'main': 5.1699250014423122,
 'part': 5.1699250014423122,
 'process': 5.1699250014423122,
 'queri': 4.1699250014423122,
 'quickli': 5.1699250014423122,
 'relev': 5.1699250014423122,
 'retriev': 3.5849625007211561,
 'scale': 5.1699250014423122,
 'search': 4.1699250014423122,
 'short': 5.1699250014423122,
 'snippet': 5.1699250014423122,
 'system': 4.1699250014423122,
 'text': 5.16992500

In [220]:
DOC1Terms = []
DOC2Terms = []
DOC3Terms = []

for i in index:        #This loop allows us to construct a bag of words for each document based on the presence of the term
    #print i
    #print index[i].sum()
    
    if index[i][0] > 0:
        DOC1Terms.append(i)
        
    if index[i][1] > 0:
        DOC2Terms.append(i)
    
    if index[i][2] > 0:
        DOC3Terms.append(i)
    

In [221]:
DOC1Terms
DOC2Terms
DOC3Terms

['allow',
 'base',
 'collect',
 'default',
 'document',
 'engine',
 'file',
 'glimps',
 'index',
 'inform',
 'larger',
 'part',
 'queri',
 'quickli',
 'retriev',
 'search',
 'system',
 'us',
 'web']

['critic',
 'document',
 'effici',
 'evalu',
 'feedback',
 'index',
 'larg',
 'main',
 'process',
 'queri',
 'relev',
 'retriev',
 'scale',
 'system',
 'updat']

['cluster',
 'creat',
 'document',
 'engine',
 'full',
 'good',
 'retriev',
 'search',
 'short',
 'snippet',
 'text',
 'web']

### Below, we are calculating the length or norm of each document

Calculating the lengths of each document involves taking the square root of the sum of the squared idf values for all terms occurring in each document

In [229]:
DOC1NormSquared = 0

for term in DOC1Terms:
    term_IDF_weight = term_Dict[term]
    #print term_IDF_weight**2
    DOC1NormSquared += term_IDF_weight**2
    
    
    
DOC1Norm = np.sqrt(DOC1NormSquared)
DOC1Norm

20.592302666166233

In [231]:
DOC2NormSquared = 0

for term in DOC2Terms:
    term_IDF_weight = term_Dict[term]
    #print term_IDF_weight**2
    DOC2NormSquared += term_IDF_weight**2
    
    
    
DOC2Norm = np.sqrt(DOC2NormSquared)
DOC2Norm

18.578212535696156

In [230]:
DOC3NormSquared = 0

for term in DOC3Terms:
    term_IDF_weight = term_Dict[term]
    #print term_IDF_weight**2
    DOC3NormSquared += term_IDF_weight**2
    
    
    
DOC3Norm = np.sqrt(DOC3NormSquared)
DOC3Norm

16.277764203350621

In [232]:
# Computing document 1 cosine similarity normalized final score:

3.421818916011/(20.592302666166233*1.0131856365479133) 

0.16400726469382268

In [233]:
# Computing document 2 cosine similarity normalized final score:

1.026541534107/(18.578212535696156*1.0131856365479133)

0.05453603684701764

In [234]:
# Computing document 3 cosine similarity normalized final score:

0.684368825059/(16.277764203350621*1.0131856365479133) 

0.04149601963812038

In [210]:
Term_IDF.iloc[1]

0     5.16993
1     5.16993
2     5.16993
3     5.16993
4     5.16993
5     5.16993
6     5.16993
7     3.58496
8     5.16993
9     4.16993
10    5.16993
11    5.16993
12    5.16993
13    5.16993
14    5.16993
15    5.16993
16    4.16993
17    5.16993
18    5.16993
19    5.16993
20    5.16993
21    5.16993
22    5.16993
23    4.16993
24    5.16993
25    5.16993
26    3.58496
27    5.16993
28    4.16993
29    5.16993
30    5.16993
31    4.16993
32    5.16993
33    5.16993
34    5.16993
35    4.16993
Name: 0, dtype: object

## 2 Indexing Models and Term Weighting

In [432]:
D = {'Index' : pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10')),
    'Term1' : pd.Series((0,5,3,1,0,2,2,3,0,1)),
     'Term2' : pd.Series((3,0,0,8,1,0,5,3,0,0)),
     'Term3' : pd.Series((1,0,4,0,0,2,0,0,3,5)),
     'Term4' : pd.Series((0,0,3,3,0,0,3,2,3,0)),
     'Term5' : pd.Series((0,3,4,0,0,0,0,0,3,2)),
     'Term6' : pd.Series((2,0,0,1,5,4,1,0,0,4)),
     'Term7' : pd.Series((1,0,0,4,4,0,4,1,0,0)),
     'Term8' : pd.Series((0,2,5,0,2,1,2,3,0,2))
    }

In [433]:
D = pd.DataFrame(data=D)
D = D.set_index('Index')

In [434]:
D

Term1  Term2  Term3  Term4  Term5  Term6  Term7  Term8
Index                                                        
Doc1       0      3      1      0      0      2      1      0
Doc2       5      0      0      0      3      0      0      2
Doc3       3      0      4      3      4      0      0      5
Doc4       1      8      0      3      0      1      4      0
Doc5       0      1      0      0      0      5      4      2
Doc6       2      0      2      0      0      4      0      1
Doc7       2      5      0      3      0      1      4      2
Doc8       3      3      0      2      0      0      1      3
Doc9       0      0      3      3      3      0      0      0
Doc10      1      0      5      0      2      4      0      2

In [435]:
DF = np.array([(D!=0).sum(0)])   #Document frequency- this is the number of documents in the collection
                                #in which a term appears

In [436]:
DF

array([[7, 5, 5, 5, 4, 6, 5, 7]])

### a) Compute the new weights for all the terms in document DOC4 using the tf x idf approach.

In [437]:
NDocs = 10

In [438]:
# Create a matrix with all entries = NDocs
NMatrix=np.ones(np.shape(D), dtype=float)*NDocs
np.set_printoptions(precision=4,suppress=True,linewidth=120)
print NMatrix

[[ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]]


In [439]:
NMatrix.shape
DF.shape

(10, 8)

(1, 8)

In [440]:
IDF = np.log2(np.divide(NMatrix, DF))

In [441]:
IDF

array([[ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146],
       [ 0.5146,  1.    ,  1.    ,  1.    ,  1.3219,  0.737 ,  1.    ,  0.5146]])

In [442]:
TFIDF = D * IDF

In [443]:
TFIDF

Term1  Term2  Term3  Term4     Term5     Term6  Term7     Term8
Index                                                                    
Doc1   0.000000    3.0    1.0    0.0  0.000000  1.473931    1.0  0.000000
Doc2   2.572866    0.0    0.0    0.0  3.965784  0.000000    0.0  1.029146
Doc3   1.543720    0.0    4.0    3.0  5.287712  0.000000    0.0  2.572866
Doc4   0.514573    8.0    0.0    3.0  0.000000  0.736966    4.0  0.000000
Doc5   0.000000    1.0    0.0    0.0  0.000000  3.684828    4.0  1.029146
Doc6   1.029146    0.0    2.0    0.0  0.000000  2.947862    0.0  0.514573
Doc7   1.029146    5.0    0.0    3.0  0.000000  0.736966    4.0  1.029146
Doc8   1.543720    3.0    0.0    2.0  0.000000  0.000000    1.0  1.543720
Doc9   0.000000    0.0    3.0    3.0  3.965784  0.000000    0.0  0.000000
Doc10  0.514573    0.0    5.0    0.0  2.643856  2.947862    0.0  1.029146

### b) Compute the new weights for all the terms in documents DOC4 using the signal-to-noise ratio approach.

In [444]:
#Here, tfk is the total number of times a term k appears in the collection. Note that this
# is different from DF
tfk = np.array([(D).sum(0)])
tfk = pd.DataFrame(tfk)
tfk

0   1   2   3   4   5   6   7
0  17  20  15  14  12  17  14  17

In [445]:
# tfik is the frequency with which a term occurs in document i in the collection, so it is 
# just the entries of dataframe D for any given document
tfik = D
tfik

Term1  Term2  Term3  Term4  Term5  Term6  Term7  Term8
Index                                                        
Doc1       0      3      1      0      0      2      1      0
Doc2       5      0      0      0      3      0      0      2
Doc3       3      0      4      3      4      0      0      5
Doc4       1      8      0      3      0      1      4      0
Doc5       0      1      0      0      0      5      4      2
Doc6       2      0      2      0      0      4      0      1
Doc7       2      5      0      3      0      1      4      2
Doc8       3      3      0      2      0      0      1      3
Doc9       0      0      3      3      3      0      0      0
Doc10      1      0      5      0      2      4      0      2

In [446]:
tfk

0   1   2   3   4   5   6   7
0  17  20  15  14  12  17  14  17

In [447]:
tfk.ix[0]

0    17
1    20
2    15
3    14
4    12
5    17
6    14
7    17
Name: 0, dtype: int64

In [451]:
TFKlist = []
for i in tfk.loc[0]:
    TFKlist.append(i)

In [452]:
TFKlist

[17, 20, 15, 14, 12, 17, 14, 17]

In [453]:
tfik['Term1']/17   #This is the array of values we should get for the pk values in Term1. Compare these to 
                    # pk below

Index
Doc1     0.000000
Doc2     0.294118
Doc3     0.176471
Doc4     0.058824
Doc5     0.000000
Doc6     0.117647
Doc7     0.117647
Doc8     0.176471
Doc9     0.000000
Doc10    0.058824
Name: Term1, dtype: float64

In [454]:
tfik.shape
tfk.shape

(10, 8)

(1, 8)

In [455]:
tfik.ix[1]

Term1    5
Term2    0
Term3    0
Term4    0
Term5    3
Term6    0
Term7    0
Term8    2
Name: Doc2, dtype: int64

In [461]:
# pk is the probability that a term occurs in document i, and is computed as tfik/tfk

pk = pd.DataFrame(tfik)
pk = tfik/TFKlist
pk

Term1     Term2     Term3     Term4     Term5     Term6     Term7  \
Index                                                                         
Doc1   0.000000  0.008824  0.003922  0.000000  0.000000  0.006920  0.004202   
Doc2   0.294118  0.000000  0.000000  0.000000  0.250000  0.000000  0.000000   
Doc3   0.176471  0.000000  0.266667  0.214286  0.333333  0.000000  0.000000   
Doc4   0.058824  0.400000  0.000000  0.214286  0.000000  0.058824  0.285714   
Doc5   0.000000  0.050000  0.000000  0.000000  0.000000  0.294118  0.285714   
Doc6   0.117647  0.000000  0.133333  0.000000  0.000000  0.235294  0.000000   
Doc7   0.117647  0.250000  0.000000  0.214286  0.000000  0.058824  0.285714   
Doc8   0.176471  0.150000  0.000000  0.142857  0.000000  0.000000  0.071429   
Doc9   0.000000  0.000000  0.200000  0.214286  0.250000  0.000000  0.000000   
Doc10  0.058824  0.000000  0.333333  0.000000  0.166667  0.235294  0.000000   

          Term8  
Index            
Doc1   0.000000  
Doc2   0.117647  
Doc3   0.294118  
Doc4   0.000000  
Doc5   0.117647  
Doc6   0.058824  
Doc7   0.117647  
Doc8   0.176471  
Doc9   0.000000  
Doc10  0.117647

In [471]:
# We compute info here as -pk log pk
INFO = -pk*(np.log(pk))

In [472]:
INFO

Term1     Term2     Term3     Term4     Term5     Term6     Term7  \
Index                                                                         
Doc1        NaN  0.041738  0.021730       NaN       NaN  0.034417  0.022993   
Doc2   0.359934       NaN       NaN       NaN  0.346574       NaN       NaN   
Doc3   0.306106       NaN  0.352468  0.330095  0.366204       NaN       NaN   
Doc4   0.166660  0.366516       NaN  0.330095       NaN  0.166660  0.357932   
Doc5        NaN  0.149787       NaN       NaN       NaN  0.359934  0.357932   
Doc6   0.251772       NaN  0.268654       NaN       NaN  0.340452       NaN   
Doc7   0.251772  0.346574       NaN  0.330095       NaN  0.166660  0.357932   
Doc8   0.306106  0.284568       NaN  0.277987       NaN       NaN  0.188504   
Doc9        NaN       NaN  0.321888  0.330095  0.346574       NaN       NaN   
Doc10  0.166660       NaN  0.366204       NaN  0.298627  0.340452       NaN   

          Term8  
Index            
Doc1        NaN  
Doc2   0.251772  
Doc3   0.359934  
Doc4        NaN  
Doc5   0.251772  
Doc6   0.166660  
Doc7   0.251772  
Doc8   0.306106  
Doc9        NaN  
Doc10  0.251772

In [475]:
# Once we have pk, we can take the average of all pk values for a given term and that is its
# average info (AVG-INFO) 
AVG_INFO = np.array([(INFO).sum(0)])
AVG_INFO

# NOISE is defined as -(AVG-INFO)
NOISE = -AVG_INFO
NOISE

# SIGNAL is defined as log(tfk)+NOISE
SIGNAL = np.log2(tfk) + NOISE
SIGNAL


array([[ 1.809 ,  1.1892,  1.3309,  1.5984,  1.358 ,  1.4086,  1.2853,  1.8398]])

array([[-1.809 , -1.1892, -1.3309, -1.5984, -1.358 , -1.4086, -1.2853, -1.8398]])

0         1         2         3         4         5         6  \
0  2.278453  3.132745  2.575947  2.208986  2.226985  2.678889  2.522061   

          7  
0  2.247673

In [493]:
SIGNAL.loc[0][1]

3.1327453651917749

In [492]:
SIGNAL_LIST = []

for i in SIGNAL.loc[0]:
    #print i
    SIGNAL_LIST.append(i)

In [494]:
SIGNAL_LIST

[2.2784525569834724,
 3.1327453651917749,
 2.575946511556277,
 2.2089862942378602,
 2.2269846457338316,
 2.6788894638100746,
 2.5220612630232093,
 2.2476732544090874]

In [497]:
# The WEIGHT for term k in document i is computed as below:

Wik = tfik * SIGNAL_LIST


Wik


Term1      Term2      Term3     Term4     Term5      Term6  \
Index                                                                   
Doc1    0.000000   0.552837   0.151526  0.000000  0.000000   0.315163   
Doc2   11.392263   0.000000   0.000000  0.000000  6.680954   0.000000   
Doc3    6.835358   0.000000  10.303786  6.626959  8.907939   0.000000   
Doc4    2.278453  25.061963   0.000000  6.626959  0.000000   2.678889   
Doc5    0.000000   3.132745   0.000000  0.000000  0.000000  13.394447   
Doc6    4.556905   0.000000   5.151893  0.000000  0.000000  10.715558   
Doc7    4.556905  15.663727   0.000000  6.626959  0.000000   2.678889   
Doc8    6.835358   9.398236   0.000000  4.417973  0.000000   0.000000   
Doc9    0.000000   0.000000   7.727840  6.626959  6.680954   0.000000   
Doc10   2.278453   0.000000  12.879733  0.000000  4.453969  10.715558   

           Term7      Term8  
Index                        
Doc1    0.148357   0.000000  
Doc2    0.000000   4.495347  
Doc3    0.000000  11.238366  
Doc4   10.088245   0.000000  
Doc5   10.088245   4.495347  
Doc6    0.000000   2.247673  
Doc7   10.088245   4.495347  
Doc8    2.522061   6.743020  
Doc9    0.000000   0.000000  
Doc10   0.000000   4.495347

### c) Using the Keyword Discrimination approach, determine if Term4 is a good index term or not (by computing its discriminant). To compute average similarities use Cosine similarity as your similarity measure. Show your work. 

In [499]:
D = {'Index' : pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10')),
    'Term1' : pd.Series((0,5,3,1,0,2,2,3,0,1)),
     'Term2' : pd.Series((3,0,0,8,1,0,5,3,0,0)),
     'Term3' : pd.Series((1,0,4,0,0,2,0,0,3,5)),
     'Term4' : pd.Series((0,0,3,3,0,0,3,2,3,0)),
     'Term5' : pd.Series((0,3,4,0,0,0,0,0,3,2)),
     'Term6' : pd.Series((2,0,0,1,5,4,1,0,0,4)),
     'Term7' : pd.Series((1,0,0,4,4,0,4,1,0,0)),
     'Term8' : pd.Series((0,2,5,0,2,1,2,3,0,2))
    }

D = pd.DataFrame(data=D)
D = D.set_index('Index')

partc = pd.DataFrame(D)
partc

Term1  Term2  Term3  Term4  Term5  Term6  Term7  Term8
Index                                                        
Doc1       0      3      1      0      0      2      1      0
Doc2       5      0      0      0      3      0      0      2
Doc3       3      0      4      3      4      0      0      5
Doc4       1      8      0      3      0      1      4      0
Doc5       0      1      0      0      0      5      4      2
Doc6       2      0      2      0      0      4      0      1
Doc7       2      5      0      3      0      1      4      2
Doc8       3      3      0      2      0      0      1      3
Doc9       0      0      3      3      3      0      0      0
Doc10      1      0      5      0      2      4      0      2

In [500]:
# Compute Dstar, the centriod document vector

Dstar = np.array([(partc).mean()])
Dstar

array([[ 1.7,  2. ,  1.5,  1.4,  1.2,  1.7,  1.4,  1.7]])

In [501]:
TOTAL_SIM = 0

Index = pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10'))

for i in Index:
    sim = (np.dot(Dstar,partc.loc[i]))/(la.norm(Dstar)*la.norm(partc.loc[i]))
    TOTAL_SIM += sim
    

AVG_SIM = TOTAL_SIM/len(partc)
AVG_SIM   

array([ 0.6806])

In [502]:
# Make a dataframe with term4 removed
removed_T4 = partc[['Term1','Term2','Term3','Term5','Term6','Term7','Term8']]
removed_T4

Term1  Term2  Term3  Term5  Term6  Term7  Term8
Index                                                 
Doc1       0      3      1      0      2      1      0
Doc2       5      0      0      3      0      0      2
Doc3       3      0      4      4      0      0      5
Doc4       1      8      0      0      1      4      0
Doc5       0      1      0      0      5      4      2
Doc6       2      0      2      0      4      0      1
Doc7       2      5      0      0      1      4      2
Doc8       3      3      0      0      0      1      3
Doc9       0      0      3      3      0      0      0
Doc10      1      0      5      2      4      0      2

In [503]:
# Here, we need to compute Dstar again with Term4 removed to get the right dimensions

Dstar4 = np.array([(removed_T4).mean()])
Dstar4

array([[ 1.7,  2. ,  1.5,  1.2,  1.7,  1.4,  1.7]])

In [504]:
TOTAL_SIM4 = 0

Index = pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10'))

for i in Index:
    sim = (np.dot(Dstar4,removed_T4.loc[i]))/(la.norm(Dstar4)*la.norm(removed_T4.loc[i]))
    TOTAL_SIM4 += sim

AVG_SIM4 = TOTAL_SIM4/len(removed_T4)
AVG_SIM4

array([ 0.6784])

### Determine if Term4 is a good index term or not by computing its discriminant

In [505]:
DISC4 = AVG_SIM4 - AVG_SIM
DISC4

array([-0.0022])

### As we can see, the discriminant of Term4 is -0.0022, so it is not a good discriminant.

## 3 Vector-Space Retrieval Model

In [ ]:
np.set_printoptions(precision=8,suppress=True,linewidth=120)

In [506]:
index=['DOC1','DOC2','DOC3','DOC4','DOC5','DOC6','DOC7','DOC8','DOC9','DOC10','QUERY']
columns = ['A','B','C','D','E','F','G','H']
table3 = pd.read_csv('HW2.csv',names=columns)

In [507]:
table3

A  B  C  D  E  F  G  H
DOC1   0  3  4  0  0  2  4  0
DOC2   5  5  0  0  4  0  4  3
DOC3   3  0  4  3  4  0  0  5
DOC4   0  7  0  3  2  0  4  3
DOC5   0  1  0  0  0  5  4  2
DOC6   2  0  2  0  0  4  0  1
DOC7   3  5  3  4  0  0  4  2
DOC8   0  3  0  0  0  4  4  2
DOC9   0  0  3  3  3  0  0  1
DOC10  0  5  0  0  0  4  4  2
QUERY  2  1  1  0  2  0  3  0

### a. Compute the ranking score for each document based on each of the following query-document similarity measures (sort the documents in decreasing order of the rank score)

In [509]:
from numpy import linalg as LA

In [510]:
# Dot product
Query = table3.loc['QUERY']


DP_rank_dict = {}

for i in index[:10]:
    print i,'dot product is:',np.dot(Query,table3.loc[i])
    rank = np.dot(Query,table3.loc[i])
    DP_rank_dict[i]=rank
    
    
import operator
sorted_dict = sorted(DP_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 dot product is: 19
DOC2 dot product is: 35
DOC3 dot product is: 18
DOC4 dot product is: 23
DOC5 dot product is: 13
DOC6 dot product is: 6
DOC7 dot product is: 26
DOC8 dot product is: 15
DOC9 dot product is: 9
DOC10 dot product is: 17


[('DOC2', 35),
 ('DOC7', 26),
 ('DOC4', 23),
 ('DOC1', 19),
 ('DOC3', 18),
 ('DOC10', 17),
 ('DOC8', 15),
 ('DOC5', 13),
 ('DOC9', 9),
 ('DOC6', 6)]

In [511]:
# Cosine similarity
Query = table3.loc['QUERY']


CS_rank_dict = {}


for i in index[:10]:
    print i,'cosine similarity is:',(np.dot(Query,table3.loc[i]))/(LA.norm(Query)*LA.norm(table3.loc[i]))
    rank = (np.dot(Query,table3.loc[i]))/(LA.norm(Query)*LA.norm(table3.loc[i]))
    CS_rank_dict[i]=rank
    
    
import operator
sorted_dict = sorted(CS_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 cosine similarity is: 0.649786289654
DOC2 cosine similarity is: 0.841725622927
DOC3 cosine similarity is: 0.476831648543
DOC4 cosine similarity is: 0.565706813135
DOC5 cosine similarity is: 0.43973155947
DOC6 cosine similarity is: 0.275298880645
DOC7 cosine similarity is: 0.671093452796
DOC8 cosine similarity is: 0.512989176043
DOC9 cosine similarity is: 0.390199486286
DOC10 cosine similarity is: 0.499352471129


[('DOC2', 0.84172562292679309),
 ('DOC7', 0.67109345279556754),
 ('DOC1', 0.64978628965393082),
 ('DOC4', 0.56570681313477411),
 ('DOC8', 0.51298917604257699),
 ('DOC10', 0.49935247112913966),
 ('DOC3', 0.4768316485434157),
 ('DOC5', 0.43973155946962988),
 ('DOC9', 0.39019948628585394),
 ('DOC6', 0.27529888064467412)]

In [513]:
# Dice's coefficient

Query = table3.loc['QUERY']


DC_rank_dict = {}


for i in index[:10]:
    numerator = 2*(np.dot(Query,table3.loc[i]))
    denominator = (LA.norm(Query)+LA.norm(table3.loc[i]))
    rank = numerator/denominator
    print i,'dice coefficient is:',rank
    DC_rank_dict[i]=rank
    
    
import operator
sorted_dict = sorted(DC_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 dice coefficient is: 3.43359959925
DOC2 dice coefficient is: 5.03659048533
DOC3 dice coefficient is: 2.76515684634
DOC4 dice coefficient is: 3.36103066234
DOC5 dice coefficient is: 2.33367433441
DOC6 dice coefficient is: 1.28220211292
DOC7 dice coefficient is: 3.9253894106
DOC8 dice coefficient is: 2.71073652572
DOC9 dice coefficient is: 1.86520735718
DOC10 dice coefficient is: 2.79395059287


[('DOC2', 5.036590485333539),
 ('DOC7', 3.925389410604927),
 ('DOC1', 3.4335995992473243),
 ('DOC4', 3.3610306623413897),
 ('DOC10', 2.7939505928676991),
 ('DOC3', 2.7651568463381007),
 ('DOC8', 2.7107365257215719),
 ('DOC5', 2.3336743344147948),
 ('DOC9', 1.8652073571770149),
 ('DOC6', 1.2822021129186529)]

In [514]:
# Jaccard's coefficient

Query = table3.loc['QUERY']


from math import*
 
def jaccard_similarity(x,y):
 
    intersection_cardinality = len(set.intersection(*[set(x), set(y)]))
    union_cardinality = len(set.union(*[set(x), set(y)]))
    return intersection_cardinality/float(union_cardinality)


JC_rank_dict = {}


for i in index[:10]:
    print i,'jaccard similarity is:',jaccard_similarity(Query,table3.loc[i])
    rank = jaccard_similarity(Query,table3.loc[i])
    JC_rank_dict[i]=rank
    
    
import operator
sorted_dict = sorted(JC_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 jaccard similarity is: 0.6
DOC2 jaccard similarity is: 0.333333333333
DOC3 jaccard similarity is: 0.333333333333
DOC4 jaccard similarity is: 0.5
DOC5 jaccard similarity is: 0.5
DOC6 jaccard similarity is: 0.6
DOC7 jaccard similarity is: 0.5
DOC8 jaccard similarity is: 0.6
DOC9 jaccard similarity is: 0.75
DOC10 jaccard similarity is: 0.333333333333


[('DOC9', 0.75),
 ('DOC8', 0.6),
 ('DOC1', 0.6),
 ('DOC6', 0.6),
 ('DOC7', 0.5),
 ('DOC4', 0.5),
 ('DOC5', 0.5),
 ('DOC2', 0.3333333333333333),
 ('DOC3', 0.3333333333333333),
 ('DOC10', 0.3333333333333333)]

### b. Compare the ranking obtained when binary term weights are used to do the ranking obtained in part a instead of raw term weights. (Do this only for dot product as the similarity measure, and explain any discrepancy between the two rankings)

In [515]:
BTW = table3

In [516]:
BTW

A  B  C  D  E  F  G  H
DOC1   0  3  4  0  0  2  4  0
DOC2   5  5  0  0  4  0  4  3
DOC3   3  0  4  3  4  0  0  5
DOC4   0  7  0  3  2  0  4  3
DOC5   0  1  0  0  0  5  4  2
DOC6   2  0  2  0  0  4  0  1
DOC7   3  5  3  4  0  0  4  2
DOC8   0  3  0  0  0  4  4  2
DOC9   0  0  3  3  3  0  0  1
DOC10  0  5  0  0  0  4  4  2
QUERY  2  1  1  0  2  0  3  0

In [517]:
BTW[BTW>1]=1

In [518]:
BTW

A  B  C  D  E  F  G  H
DOC1   0  1  1  0  0  1  1  0
DOC2   1  1  0  0  1  0  1  1
DOC3   1  0  1  1  1  0  0  1
DOC4   0  1  0  1  1  0  1  1
DOC5   0  1  0  0  0  1  1  1
DOC6   1  0  1  0  0  1  0  1
DOC7   1  1  1  1  0  0  1  1
DOC8   0  1  0  0  0  1  1  1
DOC9   0  0  1  1  1  0  0  1
DOC10  0  1  0  0  0  1  1  1
QUERY  1  1  1  0  1  0  1  0

In [519]:
# Dot product similarity with binary term weights
Query = BTW.loc['QUERY']


BTW_rank_dict = {}



for i in index[:10]:
    print i,'dot product is:',np.dot(Query,BTW.loc[i])
    rank = np.dot(Query,BTW.loc[i])
    BTW_rank_dict[i]=rank
     
import operator
sorted_dict = sorted(BTW_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 dot product is: 3
DOC2 dot product is: 4
DOC3 dot product is: 3
DOC4 dot product is: 3
DOC5 dot product is: 2
DOC6 dot product is: 2
DOC7 dot product is: 4
DOC8 dot product is: 2
DOC9 dot product is: 2
DOC10 dot product is: 2


[('DOC2', 4),
 ('DOC7', 4),
 ('DOC3', 3),
 ('DOC1', 3),
 ('DOC4', 3),
 ('DOC8', 2),
 ('DOC9', 2),
 ('DOC6', 2),
 ('DOC5', 2),
 ('DOC10', 2)]

In [520]:
import operator
DP_sorted_dict = sorted(DP_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
DP_sorted_dict


BTW_sorted_dict = sorted(BTW_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
BTW_sorted_dict

[('DOC2', 35),
 ('DOC7', 26),
 ('DOC4', 23),
 ('DOC1', 19),
 ('DOC3', 18),
 ('DOC10', 17),
 ('DOC8', 15),
 ('DOC5', 13),
 ('DOC9', 9),
 ('DOC6', 6)]

[('DOC2', 4),
 ('DOC7', 4),
 ('DOC3', 3),
 ('DOC1', 3),
 ('DOC4', 3),
 ('DOC8', 2),
 ('DOC9', 2),
 ('DOC6', 2),
 ('DOC5', 2),
 ('DOC10', 2)]

## Explain any discrepancy between the two rankings

#### The discrepancy between the two rankings can be explained by the fact that we are now using binary term weights instead of raw term weights. This means that a document that has multiple occurrences of a term will not have a higher ranking than another document having the same words but only one occurrence of each. This has the effect of normalizing the length of the documents because it doesn't matter if a document is very long and contains multiple references to the term in the query; its similarity only increases a certain amount once the term occurs. Any marginal occurrences of a term do not add anything to the similarity.

### c. Construct a similar table to the one above, but instead of raw term frequencies, compute the (non-normalized) tf-idf weights for the terms. Then compute the ranking scores using the Cosine Similarity. Explain any significant differences between the ranking you obtained here and the Cosine ranking from the previous part.

In [521]:
index=['DOC1','DOC2','DOC3','DOC4','DOC5','DOC6','DOC7','DOC8','DOC9','DOC10','QUERY']
columns = ['A','B','C','D','E','F','G','H']
table3c = pd.read_csv('HW2.csv',names=columns)

In [522]:
table3c

A  B  C  D  E  F  G  H
DOC1   0  3  4  0  0  2  4  0
DOC2   5  5  0  0  4  0  4  3
DOC3   3  0  4  3  4  0  0  5
DOC4   0  7  0  3  2  0  4  3
DOC5   0  1  0  0  0  5  4  2
DOC6   2  0  2  0  0  4  0  1
DOC7   3  5  3  4  0  0  4  2
DOC8   0  3  0  0  0  4  4  2
DOC9   0  0  3  3  3  0  0  1
DOC10  0  5  0  0  0  4  4  2
QUERY  2  1  1  0  2  0  3  0

In [523]:
# TFIDF

NDocs = 10


DF = np.array([(table3c!=0).sum(0)])
DF


# Create a matrix with all entries = NDocs
NMatrix=np.ones(np.shape(table3c), dtype=float)*NDocs
np.set_printoptions(precision=4,suppress=True,linewidth=120)
print NMatrix



IDF = np.log2(np.divide(NMatrix, DF))

IDF

array([[5, 8, 6, 4, 5, 5, 8, 9]])

[[ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]
 [ 10.  10.  10.  10.  10.  10.  10.  10.]]


array([[ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ],
       [ 1.    ,  0.3219,  0.737 ,  1.3219,  1.    ,  1.    ,  0.3219,  0.152 ]])

In [524]:
TFIDF = table3c * IDF
TFIDF

A         B         C         D    E    F         G         H
DOC1   0.0  0.965784  2.947862  0.000000  0.0  2.0  1.287712  0.000000
DOC2   5.0  1.609640  0.000000  0.000000  4.0  0.0  1.287712  0.456009
DOC3   3.0  0.000000  2.947862  3.965784  4.0  0.0  0.000000  0.760015
DOC4   0.0  2.253497  0.000000  3.965784  2.0  0.0  1.287712  0.456009
DOC5   0.0  0.321928  0.000000  0.000000  0.0  5.0  1.287712  0.304006
DOC6   2.0  0.000000  1.473931  0.000000  0.0  4.0  0.000000  0.152003
DOC7   3.0  1.609640  2.210897  5.287712  0.0  0.0  1.287712  0.304006
DOC8   0.0  0.965784  0.000000  0.000000  0.0  4.0  1.287712  0.304006
DOC9   0.0  0.000000  2.210897  3.965784  3.0  0.0  0.000000  0.152003
DOC10  0.0  1.609640  0.000000  0.000000  0.0  4.0  1.287712  0.304006
QUERY  2.0  0.321928  0.736966  0.000000  2.0  0.0  0.965784  0.000000

In [525]:
table3c

A  B  C  D  E  F  G  H
DOC1   0  3  4  0  0  2  4  0
DOC2   5  5  0  0  4  0  4  3
DOC3   3  0  4  3  4  0  0  5
DOC4   0  7  0  3  2  0  4  3
DOC5   0  1  0  0  0  5  4  2
DOC6   2  0  2  0  0  4  0  1
DOC7   3  5  3  4  0  0  4  2
DOC8   0  3  0  0  0  4  4  2
DOC9   0  0  3  3  3  0  0  1
DOC10  0  5  0  0  0  4  4  2
QUERY  2  1  1  0  2  0  3  0

In [526]:
# Cosine similarity
Query = table3c.loc['QUERY']


CS3_rank_dict = {}


for i in index[:10]:
    print i,'cosine similarity is:',(np.dot(Query,table3c.loc[i]))/(LA.norm(Query)*LA.norm(table3c.loc[i]))
    rank = (np.dot(Query,table3.loc[i]))/(LA.norm(Query)*LA.norm(table3.loc[i]))
    CS3_rank_dict[i]=rank
    
    
import operator
sorted_dict = sorted(CS3_rank_dict.items(), key=operator.itemgetter(1),reverse=True)
sorted_dict

DOC1 cosine similarity is: 0.649786289654
DOC2 cosine similarity is: 0.841725622927
DOC3 cosine similarity is: 0.476831648543
DOC4 cosine similarity is: 0.565706813135
DOC5 cosine similarity is: 0.43973155947
DOC6 cosine similarity is: 0.275298880645
DOC7 cosine similarity is: 0.671093452796
DOC8 cosine similarity is: 0.512989176043
DOC9 cosine similarity is: 0.390199486286
DOC10 cosine similarity is: 0.499352471129


[('DOC2', 0.82078268166812318),
 ('DOC7', 0.65561006810718581),
 ('DOC4', 0.61558701125109239),
 ('DOC1', 0.57353933467640439),
 ('DOC3', 0.51298917604257699),
 ('DOC8', 0.45883146774112349),
 ('DOC5', 0.45883146774112349),
 ('DOC10', 0.45883146774112349),
 ('DOC9', 0.34412360080584259),
 ('DOC6', 0.34412360080584259)]

## Explain any significant differences between the ranking you obtained here and the Cosine ranking from the previous part.

#### The difference between the rankings we obtained here and the rankings from part a) using Cosine similarity is that we used tf-idf weights instead of raw term frequencies. What this means is that we are making a correction for terms that occur frequently within certain documents but infrequently over the span of the entire document collection. This will give higher weights to terms that are more specific to our query or specific subject matter and lower weights to terms that simply have a high overall frequency or probability of occurrence. 

#### The importance of using this tf-idf weighting is in its power to discriminate against terms that do not really add much distiniguishing power. What I mean by this is that users will not receive much value by retrieving documents that have common words such as (the, a, it, he, she, by, yes, no, etc.) . The users will benefit more from a retrieval system that specifically grabs documents with words that have a high prbability of occurring in relevant documents but low probability in nonrelevant documents. The use of tf-idf weighting is like using stop word removal in preprocessing, because it focuses on words that are not likely to be found in all documents. It could be said that tf-idf weighting is a way of reducing the noise in the dataset and focusing more on the signal.

## 4. Probabilistic Retrieval Model

In [546]:
np.set_printoptions(precision=8,suppress=True,linewidth=120)

In [568]:
P4 = {'Index' : pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10')),
    'T1' : pd.Series((1,0,1,0,1,1,0,0,1,1)),
     'T2' : pd.Series((0,1,0,1,1,1,0,0,1,0)),
     'T3' : pd.Series((1,0,1,1,0,0,0,1,1,0)),
     'T4' : pd.Series((1,1,1,0,1,1,0,1,0,1)),
     'T5' : pd.Series((0,0,1,1,0,1,0,1,1,1)),
     'T6' : pd.Series((0,1,0,1,0,1,1,0,1,0)),
      'Relevance' : pd.Series(('R','R','NR','NR','NR','NR','R','NR','R','NR'))
    }

In [569]:
P4 = pd.DataFrame(P4,columns=['Index','T1','T2','T3','T4','T5','T6','Relevance'])

In [570]:
P4 = P4.set_index('Index')

In [571]:
P4 

T1  T2  T3  T4  T5  T6 Relevance
Index                                  
Doc1    1   0   1   1   0   0         R
Doc2    0   1   0   1   0   1         R
Doc3    1   0   1   1   1   0        NR
Doc4    0   1   1   0   1   1        NR
Doc5    1   1   0   1   0   0        NR
Doc6    1   1   0   1   1   1        NR
Doc7    0   0   0   0   0   1         R
Doc8    0   0   1   1   1   0        NR
Doc9    1   1   1   0   1   1         R
Doc10   1   0   0   1   1   0        NR

In [572]:
Relevant = P4[P4['Relevance']=='R']
Relevant


# Compute Pi, the probability that a term i occurs in a relevant document
Pi = np.array([(Relevant==1).sum(0)])/len(Relevant)
Pi = Pi[0][0:6]
Pi

T1  T2  T3  T4  T5  T6 Relevance
Index                                  
Doc1    1   0   1   1   0   0         R
Doc2    0   1   0   1   0   1         R
Doc7    0   0   0   0   0   1         R
Doc9    1   1   1   0   1   1         R

array([ 0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.25,  0.75])

In [573]:
NonRelevant = P4[P4['Relevance']=='NR']
NonRelevant


# Compute Qi, the probability that a term i occurs in a nonrelevant document
Qi = np.array([(NonRelevant==1).sum(0)])/len(NonRelevant)
Qi = Qi[0][0:6]
Qi

T1  T2  T3  T4  T5  T6 Relevance
Index                                  
Doc3    1   0   1   1   1   0        NR
Doc4    0   1   1   0   1   1        NR
Doc5    1   1   0   1   0   0        NR
Doc6    1   1   0   1   1   1        NR
Doc8    0   0   1   1   1   0        NR
Doc10   1   0   0   1   1   0        NR

array([ 0.66666667,  0.5       ,  0.5       ,  0.83333333,  0.83333333,  0.33333333])

In [574]:
# Calculate D*, the centroid vector

Dstar = np.array([(P4).mean()])
Dstar

array([[ 0.6,  0.5,  0.5,  0.7,  0.6,  0.5]])

In [575]:
P4values = P4[['T1','T2','T3','T4','T5','T6']]
P4values

T1  T2  T3  T4  T5  T6
Index                        
Doc1    1   0   1   1   0   0
Doc2    0   1   0   1   0   1
Doc3    1   0   1   1   1   0
Doc4    0   1   1   0   1   1
Doc5    1   1   0   1   0   0
Doc6    1   1   0   1   1   1
Doc7    0   0   0   0   0   1
Doc8    0   0   1   1   1   0
Doc9    1   1   1   0   1   1
Doc10   1   0   0   1   1   0

In [576]:
TOTAL_SIM = 0

Index = pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10'))

for i in Index:
    sim = (np.dot(Dstar,P4values.loc[i]))/(LA.norm(Dstar)*LA.norm(P4values.loc[i]))
    TOTAL_SIM += sim

In [577]:
AVG_SIM = TOTAL_SIM/len(P4values)
AVG_SIM

array([ 0.74646775])

In [578]:
Query4 = [1,1,0,1,0,1]  #What do we do with this query to compute the discriminants?

In [579]:
D11 = [0,1,1,0,0,1]
D12 = [1,0,1,1,0,1]

In [580]:
# Construct a new dataframe with D11:

D11 = {'Index' : pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10','Doc11')),
    'T1' : pd.Series((1,0,1,0,1,1,0,0,1,1,0)),
     'T2' : pd.Series((0,1,0,1,1,1,0,0,1,0,1)),
     'T3' : pd.Series((1,0,1,1,0,0,0,1,1,0,1)),
     'T4' : pd.Series((1,1,1,0,1,1,0,1,0,1,0)),
     'T5' : pd.Series((0,0,1,1,0,1,0,1,1,1,0)),
     'T6' : pd.Series((0,1,0,1,0,1,1,0,1,0,1)),
    }

D11 = pd.DataFrame(D11,columns=['Index','T1','T2','T3','T4','T5','T6'])

D11 = D11.set_index('Index')

D11

T1  T2  T3  T4  T5  T6
Index                        
Doc1    1   0   1   1   0   0
Doc2    0   1   0   1   0   1
Doc3    1   0   1   1   1   0
Doc4    0   1   1   0   1   1
Doc5    1   1   0   1   0   0
Doc6    1   1   0   1   1   1
Doc7    0   0   0   0   0   1
Doc8    0   0   1   1   1   0
Doc9    1   1   1   0   1   1
Doc10   1   0   0   1   1   0
Doc11   0   1   1   0   0   1

In [581]:
TOTAL_SIMD11 = 0

IndexD11 = pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10','Doc11'))

for i in IndexD11:
    sim = (np.dot(Dstar,D11.loc[i]))/(LA.norm(Dstar)*LA.norm(D11.loc[i]))
    TOTAL_SIMD11 += sim

In [582]:
AVG_SIMD11 = TOTAL_SIMD11/len(D11)
AVG_SIMD11

array([ 0.73484247])

In [583]:
DiscQD11 = AVG_SIMD11-AVG_SIM 
DiscQD11

array([-0.01162529])

In [584]:
# Construct a new dataframe with D12:

D12 = {'Index' : pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10','Doc12')),
    'T1' : pd.Series((1,0,1,0,1,1,0,0,1,1,1)),
     'T2' : pd.Series((0,1,0,1,1,1,0,0,1,0,0)),
     'T3' : pd.Series((1,0,1,1,0,0,0,1,1,0,1)),
     'T4' : pd.Series((1,1,1,0,1,1,0,1,0,1,1)),
     'T5' : pd.Series((0,0,1,1,0,1,0,1,1,1,0)),
     'T6' : pd.Series((0,1,0,1,0,1,1,0,1,0,1)),
    }

D12 = pd.DataFrame(D12,columns=['Index','T1','T2','T3','T4','T5','T6'])

D12 = D12.set_index('Index')

D12

T1  T2  T3  T4  T5  T6
Index                        
Doc1    1   0   1   1   0   0
Doc2    0   1   0   1   0   1
Doc3    1   0   1   1   1   0
Doc4    0   1   1   0   1   1
Doc5    1   1   0   1   0   0
Doc6    1   1   0   1   1   1
Doc7    0   0   0   0   0   1
Doc8    0   0   1   1   1   0
Doc9    1   1   1   0   1   1
Doc10   1   0   0   1   1   0
Doc12   1   0   1   1   0   1

In [585]:
TOTAL_SIMD12 = 0

IndexD12 = pd.Series(('Doc1','Doc2','Doc3','Doc4','Doc5','Doc6','Doc7','Doc8','Doc9','Doc10','Doc12'))

for i in IndexD12:
    sim = (np.dot(Dstar,D12.loc[i]))/(LA.norm(Dstar)*LA.norm(D12.loc[i]))
    TOTAL_SIMD12 += sim

In [586]:
AVG_SIMD12 = TOTAL_SIMD12/len(D12)
AVG_SIMD12

array([ 0.75328237])

In [587]:
DiscQd12 = AVG_SIMD12-AVG_SIM 
DiscQd12

array([ 0.00681462])

## Based on the discriminants, should these documents be retrieved? Explain your answer.

### The discriminant for document 11 is -0.01162529, so we could say it is a poor discriminant because it is negative and close to zero;  we should not retrieve it.

### The discriminant for document 12 is 0.00681462, so we could say it is a good discriminant because it is positive and therefore we should  retrieve it. We should note, however, that we prefer discriminant values that are large and this one is somewhat close to zero.

### A discriminant is considered good if the similarity among documents changes significantly when the discriminating term (or in this case document) is removed. Both of these documents, D11 and D12, do not change the average similarity much, so there is not a very strong reason for us to retrieve these documents.

### See word document for the answer to question #5